specs/error_state/ErrorOutOfGasDynamicMemoryExpansion.md

```diff
@@ -0,0 +1,59 @@
+# ErrorOutOfGasDynamicMemoryExpansion state
+
+## Procedure
+
+Handle the corresponding out of gas errors for `CREATE`, `RETURN` and `REVERT` opcodes.
+
+### EVM behavior
+
+For the current `go-ethereum` code, the out of gas error may occur for `constant gas` or `dynamic gas`. The gas cost is calculated as:
+
+```
+gas_cost = constant_gas + dynamic_gas
+```
+
+The constant gas is different for `CREATE`, `RETURN` and `REVERT`.
+
+```
+if opcode == CREATE:
+    constant_gas = 32000
+else:
+    constant_gas = 0
+```
+
+They are also same for dynamic gas calculation. The dynamic gas is calculated as:
+
+```
+dynamic_gas = memory_expansion_cost
+```
+
+The memory expansion gas cost is calculated as:
+
+```
+next_memory_word_size = (destination_offset + copy_byte_size + 31) // 32
+# `current_memory_word_size` is fetched from the execution step.
+if next_memory_word_size <= current_memory_word_size:
+    memory_expansion_gas_cost = 0
+else:
+    memory_expansion_gas_cost = (
+        3 * (next_memory_word_size - current_memory_word_size)
+        + next_memory_word_size * next_memory_word_size // 512
+        - current_memory_word_size * current_memory_word_size // 512
+    )
+```
+
+### Constraints
+
+1. Current opcode is one of `CREATE`, `RETURN` and `REVERT`.
+2. Constrain `gas_left < gas_cost`.
+3. Current call must fail.
+4. If it's a root call, it transits to `EndTx`.
+5. If it isn't a root call, it restores caller's context by reading to `rw_table`, then does step state transition to it.
+6. Constrain `rw_counter_end_of_reversion = rw_counter_end_of_step + reversible_counter`.
+
+### Lookups
+
+1. `3` stack pop for `CREATE` and `2` stack pop for `RETURN` and `REVERT`.
```

Comment:
Wondering whether for `CREATE` there should be a push too? Because it seems there is always a push to the stack, although sometimes an empty value:

```
if interpreter.evm.chainRules.IsHomestead && suberr == ErrCodeStoreOutOfGas {
	stackvalue.Clear()
} else if suberr != nil && suberr != ErrCodeStoreOutOfGas {
	stackvalue.Clear()
} else {
	stackvalue.SetBytes(addr.Bytes())
}
scope.Stack.push(&stackvalue)
```

Reply:
Based on my observations from the unit tests, it appears that the execution is terminated, hence there's no subsequent execution step (exec_step).

Reply:
@lightsing Sorry, I completely missed you answer. However, I am not totally sure about it as there could probably be other operations in the function of the smart contract after the contract is instantiated? If there is not enough gas for creating the new contract then the code is not set at the address and the error is returned (in `evm.go`)

```
createDataGas := uint64(len(ret)) * params.CreateDataGas
		if contract.UseGas(createDataGas) {
			evm.StateDB.SetCode(address, ret)
		} else {
			err = ErrCodeStoreOutOfGas
		}
```

and then `contract.UseGas` is not called. Let me check a bit more.

Can you point me to the tests you were observing?

Reply:
I am not sure if I am correct, but it seems to me there can be subsequent steps in the transaction. For example having (this uses Create2, but the situation is the same):

```
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.14;

contract DeployWithCreate2 {
    address public owner;
    constructor(address _owner) {
        owner = _owner;
    }
}

contract Create2Factory {
    event Deploy(address addr);
    string public message;

    function deploy(uint _salt) external {
        DeployWithCreate2 _contract = new DeployWithCreate2{
            salt: bytes32(_salt)    // the number of salt determines the address of the contract that will be deployed
        }(msg.sender);
        emit Deploy(address(_contract));

        message = "foo";
    }
}
```

Here, there is `sstore` after `create2` and I think in the case of OOG error, the `contract.Gas` is not subtracted in `evm.go/create` and error is not returned [here](https://github.com/ethereum/go-ethereum/blob/master/core/vm/instructions.go#L659), so the execution continues.

However, I couldn’t reproduce it in Remix as it seems that it ignores the gas limit setting.

Reply:
Given the following test case:
code_a:
```
PUSH1(0x00) // retLength
PUSH1(0x00) // retOffset
PUSH32(0x00) // argsLength
PUSH32(0x00) // argsOffset
PUSH1(0x00) // value
PUSH32(code_b) // addr
PUSH32(0xFFFF) // gas
CALL
STOP
```
code_b:
```
PUSH32(0xff)
PUSH32(0xffffffff)
PUSH8(0)
CREATE
STOP
```
the code_a calls code_b and results in the following traces:
```
Step { pc: 0x0000, op: PUSH1, gas: 1027575, gas_cost: 3, depth: 1, error: None, stack: [], storage: {} }
Step { pc: 0x0002, op: PUSH1, gas: 1027572, gas_cost: 3, depth: 1, error: None, stack: [0x0], storage: {} }
Step { pc: 0x0004, op: PUSH32, gas: 1027569, gas_cost: 3, depth: 1, error: None, stack: [0x0, 0x0], storage: {} }
Step { pc: 0x0025, op: PUSH32, gas: 1027566, gas_cost: 3, depth: 1, error: None, stack: [0x0, 0x0, 0x0], storage: {} }
Step { pc: 0x0046, op: PUSH1, gas: 1027563, gas_cost: 3, depth: 1, error: None, stack: [0x0, 0x0, 0x0, 0x0], storage: {} }
Step { pc: 0x0048, op: PUSH32, gas: 1027560, gas_cost: 3, depth: 1, error: None, stack: [0x0, 0x0, 0x0, 0x0, 0x0], storage: {} }
Step { pc: 0x0069, op: PUSH32, gas: 1027557, gas_cost: 3, depth: 1, error: None, stack: [0x0, 0x0, 0x0, 0x0, 0x0, 0xcafe222], storage: {} }
Step { pc: 0x008a, op: CALL, gas: 1027554, gas_cost: 68135, depth: 1, error: None, stack: [0x0, 0x0, 0x0, 0x0, 0x0, 0xcafe222, 0xffff], storage: {} }
Step { pc: 0x0000, op: PUSH32, gas: 65535, gas_cost: 3, depth: 2, error: None, stack: [], storage: {} }
Step { pc: 0x0021, op: PUSH32, gas: 65532, gas_cost: 3, depth: 2, error: None, stack: [0xff], storage: {} }
Step { pc: 0x0042, op: PUSH8, gas: 65529, gas_cost: 3, depth: 2, error: None, stack: [0xff, 0xffffffff], storage: {} }
Step { pc: 0x004b, op: CREATE, gas: 65526, gas_cost: 35184778968344, depth: 2, error: Some("out of gas"), stack: [0xff, 0xffffffff, 0x0], storage: {} }
Step { pc: 0x008b, op: STOP, gas: 959419, gas_cost: 0, depth: 1, error: None, stack: [0x0], storage: {} }]
```
Note, as soon as the oog caused by `CREATE`, the callee reverts to caller (see in depth). So there's no way for subsequent opcodes to use the callee's stack.

Reply:
Thanks! Can you point me to these tests? I would like to run them to get a better understanding.

Reply:
Here's you are. https://github.com/scroll-tech/zkevm-circuits/blob/develop/zkevm-circuits/src/evm_circuit/execution/error_oog_dynamic_memory.rs#L126
You might want to comment out L214 (skip the circuit part) and print the geth traces in ctx.